### ISSUE TITLE
Allow Salmon to receive and pass external IDs

### ISSUE THREAD

**Issue by crcox** [stsievert/salmon#117]
In the experimental psychology setting, we often construct a battery of tasks/questionnaires that we would like a participant to complete. The triplet similarity judgment task might be sandwiched between other tasks or questionnaires. This may be a model experimental researchers will adopt in fields other than Psychology.

A simple solution would involve slightly extending the API. When arriving at the query page, the url *may* include a query string defining the `ext_id` variable. The `ext_id` could be stored alongside the `puid` that is generated by Salmon.

Then, we could create a new `GET` hook that would allow the `puid` to be requested, given a valid `ext_id`. (Maybe this could also return the number of queries presented to that `ext_id`/`puid`.)

Thus, a participant might be sent to complete the triplet task served by Salmon, along with an ID that allows their responses tracked by Salmon to be linked back to their prior responses. The `GET` hook would register the `{ext_id: ext_id, salmon_id: puid}` with the next task or questionnaire.

Perhaps there is a better design for accomplishing this, but allowing Salmon to receive and store and `ext_id` would be valuable.

**Comment by crcox** [comment]
EDIT: I realize now that the "query string" comment was not technically correct. I guess it should follow the model:

```
@app.get("/{ext_id}", tags=["public"])                                                                                                                                               
async def get_query_page(request: Request, ext_id: str):
# the code, modified to do something with ext_id
```

Or, is it possible to leave `get_query_page` alone, and write a new `GET` hook, like:

```
@app.get("/register_ext_id/{ext_id}", tags=["public"])                                                                                                                                               
async def register_external_id(request: Request, ext_id: str):
# the code, doing whatever "register external_id" requires, and then
    return get_query_page(request)
```

That would probably be better, since it would leave the basic interface exactly as it is, and allow passing the `ext_id` as an option. Is a solution of this kind possible?

**Comment by stsievert** [comment]
Would allowing Salmon's participant ID to take custom values resolve this issue?

**Comment by crcox** [comment]
It probably would, unless it is important to distinguish between participants and sessions.

By which I mean: if Salmon's `puid` can be overwritten, there is the opportunity for queries issued during different sessions to be associated with a common identifier. Maybe this doesn't matter, but in general there might be a reason to distinguish between participants and sessions. The `puid` could be guaranteed unique for every unique `GET` request triggering `get_query_page`, and the `ext_id` could be whatever the experimenter etc. wants it to be.

If it is trivial to add another field to the db, then my guess is that maintaining and internal and an external id might be more flexible and could have important utility.

That said, for my current use case, allowing the `puid` to be manually specified would meet my needs.

**Comment by crcox** [comment]
PS: I'll add that the primary reason why linking these things is important to me is because my study team will be monitoring the responses for bot-like behavior. If someone is behaving badly, I'll want to exclude their data---all of it, across data collection platforms.

In general, I might be interested in carving out responses from different participant subsets post-hoc. If I knew I wanted to study groups separately a priori I would set up two instances of Salmon to ensure that responses were collected and separated by group (group code could be embedded in `ext_id`.

**Comment by stsievert** [comment]
I think this is closed by #118, which allows specifying the participant ID by hitting the endpoint `/query?puid=foobar`. That means the experimentalist can customize the user ID.

@crcox, pleas reopen if it does not resolve the issue for your study team (that I think is expressed in https://github.com/stsievert/salmon/issues/117#issuecomment-880878657).
